Source/WebCore/rendering/RenderGrid.cpp

```diff
@@ -779,6 +786,16 @@ void RenderGrid::placeItemsOnGrid(GridTrackSizingAlgorithm& algorithm, std::opti
 
     grid.setNeedsItemsPlacement(false);
 
+    if (needsMasonryLayout) {
+        HashMap<RenderBox*, GridArea> firstTrackItems;
+        Vector<RenderBox*> itemsWithDefiniteGridAxisPosition;
+        Vector<RenderBox*> itemsWithIndefinitePosition;
+
+        collectMasonryItems(grid, gridAxisTracksBeforeAutoPlacement, masonryAxisDirection, firstTrackItems, itemsWithDefiniteGridAxisPosition, itemsWithIndefinitePosition);
+        grid.setupGridForMasonryLayout();
```

Comment:
If this is going to clear `m_grid` and `m_gridItemArea`, is it necessary to run the non-masonry placement logic beforehand?

Reply:
Yes I agree that running the entire placement beforehand is probably not ideal. I did it this way to basically get an initial implementation/proof of concept done. 

I think we will need to run some form the non-masonry placement logic in order to figure out which items were placed on the first implicit line. It seems like the ideal scenario would be to figure out when we should short-circuit the standard grid placement algorithm.

Reply:
Should I open another bug report for this issue?

Reply:
Yes, can be improved in another bug

Reply:
https://bugs.webkit.org/show_bug.cgi?id=248291